File: src/main/java/xyz/kazuthecat/coffeebot/settings/SettingsAbstract.java
package xyz.kazuthecat.coffeebot.settings;

import com.google.gson.Gson;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.User;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * This class contains all the main setting functions shared by it's subclasses. It has only one abstract method
 * called writeJSON which is dictates how settings are saved, while loading of settings from this medium is handled
 * by the subclasses constructor methods.
 */
public abstract class SettingsAbstract {
    final Gson gson = new Gson();
    private final Map<String, String> defaultSettings = new HashMap<>();
    private Map<String, Boolean> userChangeable = new HashMap<>();
    private Map<String, Boolean> adminChangeable = new HashMap<>();
    Map<String, Map<String, String>> customSettings = new HashMap<>();

    /**
     * This method dictates how settings are saved for long-term storage. For SettingsDB this means writing the setting
     * to a MySQL database server and for SettingsFile this means writing all settings to disk in JSON format.
     *
     * The parameters are not used by all subclasses. Since SettingsFile dumps all settings into a file the arguments
     * passed to the function are ignored, however SettingsDB will only write one setting at a time. Including the
     * parameters for this method allows for maximum flexibility in the subclasses on how exactly to save the data.
     * @param settingName The name of the setting that's affected.
     * @param id The ID of the user or guild whose settings are being changed (or DEFAULT for owner settings).
     * @param value The new value for the setting.
     */
    abstract void writeJSON(String settingName, String id, String value);

    /**
     * This function is called by each command that wants to add it's settings. Each call creates
     * an entry in defaultSettings, and in customSettings if there were none at initialization.
     * Finally, userChangeable and adminChangeable are set.
     * @param settingName The name of the setting, e.g. something like hello.response.
     * @param defaultValue The default value for the setting (unless one was loaded previously).
     * @param userChangeable Boolean saying whether users are allowed to change this setting.
     * @param adminChangeable Boolean saying whether server admins are allowed to change this setting.
     */
    public void setDefaults(String settingName, String defaultValue, boolean userChangeable, boolean adminChangeable) {
        defaultSettings.put(settingName, defaultValue);
        customSettings.computeIfAbsent(settingName, k -> new HashMap<>());
        this.userChangeable.put(settingName, userChangeable);
        this.adminChangeable.put(settingName, adminChangeable);
    }

    /**
     * Method for changing a setting for an individual user.
     * @param user The User object for the user.
     * @param settingName The name of the setting.
     * @param value The new value for the setting.
     * @return A SettingEnum indicating whether the operation was successful.
     */
    public SettingEnum putSetting(User user, String settingName, String value) {
        if (!userChangeable.get(settingName)) {
            return SettingEnum.FORBIDDEN;
        } else {
            return putSetting(user.getId(), settingName, value);
        }
    }

    /**
     * Method for changing a setting for a specific guild.
     * @param guild The Guild object for the guild.
     * @param settingName The name of the setting.
     * @param value The new value for the setting.
     * @return A SettingEnum indicating whether the operation was successful.
     */
    public SettingEnum putSetting(Guild guild, String settingName, String value) {
        if (!userChangeable.get(settingName)) {
            return SettingEnum.FORBIDDEN;
        } else {
            return putSetting(guild.getId(), settingName, value);
        }
    }

    /**
     * Method for setting a new default value for a setting. This value will be used in all servers as long
     * as users don't have their own settings or admins have set a guild setting.
     * @param settingName The name of the setting.
     * @param value The new value for the setting.
     * @return A SettingEnum indicating whether the operation was successful.
     */
    public SettingEnum putSetting(String settingName, String value) {
        return putSetting("DEFAULT", settingName, value);
    }

    /**
     * Helper method called by the other putSetting functions.
     * This does all the "heavy lifting" when setting a new setting.
     * @param id The ID of the user or guild whose settings are being changed (or "DEFAULT" for owner settings).
     * @param settingName The name of the setting that's being changed.
     * @param value The new value for the setting.
     * @return A SettingEnum indicating if the operation was successful.
     */
    private SettingEnum putSetting(String id, String settingName, String value) {
        if (!defaultSettings.containsKey(settingName)) {
            return SettingEnum.DOESNOTEXIST;
        } else {
            customSettings.get(settingName).put(id, value);
            writeJSON(settingName, id, value);
            return SettingEnum.SUCCESSFUL;
        }
    }

    /**
     * Method for resetting a setting for an individual user.
     * @param user The User object for the user.
     * @param settingName The name of the setting.
     * @return A SettingEnum indicating whether the operation was successful.
     */
    public SettingEnum removeSetting(User user, String settingName) {
        if (!userChangeable.get(settingName)) {
            return SettingEnum.FORBIDDEN;
        } else {
            return removeSetting(user.getId(), settingName);
        }
    }

    /**
     * Method for resetting a setting for a specific guild.
     * @param guild The Guild object for the guild.
     * @param settingName The name of the setting.
     * @return A SettingEnum indicating whether the operation was successful.
     */
    public SettingEnum removeSetting(Guild guild, String settingName) {
        if (!userChangeable.get(settingName)) {
            return SettingEnum.FORBIDDEN;
        } else {
            return removeSetting(guild.getId(), settingName);
        }
    }

    /**
     * Method for resetting a changed default value to the default provided by the command itself.
     * @param settingName The name of the setting.
     * @return A SettingEnum indicating whether the operation was successful.
     */
    public SettingEnum removeSetting(String settingName) {
        return removeSetting("DEFAULT", settingName);
    }

    /**
     * Helped method used by the other removeSetting functions to change the setting.
     * @param id The ID of the user or guild whose setting is being removed (or DEFAULT for owner settings).
     * @param settingName The name of the setting that's being removed.
     * @return A SettingEnum indicating if the operation was successful.
     */
    private SettingEnum removeSetting(String id, String settingName) {
        if (!defaultSettings.containsKey(settingName)) {
            return SettingEnum.DOESNOTEXIST;
        } else {
            Map<String, String> setting = customSettings.get(settingName);
            if (setting.containsKey(id)) {
                customSettings.get(settingName).remove(id);
                writeJSON(settingName, id, null);
                return SettingEnum.SUCCESSFUL;
            } else {
                return SettingEnum.NOTSET;
            }
        }
    }

    /**
     * Method for retrieving the most relevant setting for responding to a message.
     * If a user setting is set (and the setting is currently userChangeable), that setting is returned.
     * If a guild setting is set (and the setting is currently adminChangeable), that setting is returned.
     * If the default setting has been altered, that setting is returned.
     * If the setting has not been altered, the default setting is returned.
     * @param settingName The name of the setting.
     * @param message The message requesting the setting.
     * @return A string containing the setting.
     */
    public String getSetting(String settingName, Message message) {
        String botSetting = null, userSetting = null, guildSetting = null, result;
        Map<String, String> setting = customSettings.get(settingName);

        if (setting != null) {
            if (userChangeable.get(settingName))
                userSetting = setting.get(message.getAuthor().getId());
            if (adminChangeable.get(settingName) && message.isFromGuild())
                guildSetting = setting.get(message.getGuild().getId());
            botSetting = setting.get("DEFAULT");
        }

        if (userSetting != null) {
            result = standardReplacements(userSetting, message);
        } else if (guildSetting != null) {
            result = standardReplacements(guildSetting, message);
        } else if (botSetting != null) {
            result = standardReplacements(botSetting, message);
        } else {
            result = standardReplacements(defaultSettings.get(settingName), message);
        }

        return result;
    }

    /**
     * Method for carrying out a few standard replacements on a retrieved setting.
     * Placeholders are indicated with {brackets}.
     * @param input The string we're doing replacements on.
     * @param message The message requesting the setting.
     * @return A modified version of input with all replacements carried out.
     */
    private String standardReplacements(String input, Message message) {
        return input
                .replaceAll("\\{user\\}", message.getAuthor().getAsMention())
                .replaceAll("\\{botname\\}", message.getJDA().getSelfUser().getAsMention())
                .replaceAll("\\{content\\}", message.getContentRaw().strip());
    }

    /**
     * Method for searching through the settings by looking for setting names containing the provided substring.
     * @param substring The substring we're filtering settings based on.
     * @return A set of setting names (String) whose names contain the provided substring.
     */
    public Set<String> allSettingsContaining(String substring) {
        return customSettings
                .keySet()
                .stream()
                .filter(x -> x.contains(substring))
                .collect(Collectors.toSet());
    }
}

File: src/main/kotlin/settings/SettingsAbstract.kt
package xyz.kazuthecat.coffeebot.settings

import net.dv8tion.jda.api.entities.*
import java.util.HashMap
import java.util.stream.Collectors

abstract class SettingsAbstract {
  private val defaultSettings = HashMap<String, String>()
  private val userChangeable = HashMap<String, Boolean>()
  private val adminChangeable = HashMap<String, Boolean>()
  internal var customSettings = HashMap<String, MutableMap<String, String>>()

  internal abstract fun writeJSON(settingName: String, id: String, value: String?)

  fun setDefaults(settingName: String, defaultValue: String, userChangeable: Boolean, adminChangeable: Boolean) {
    defaultSettings[settingName] = defaultValue
    this.userChangeable[settingName] = userChangeable
    this.adminChangeable[settingName] = adminChangeable
    customSettings.computeIfAbsent(settingName) { _ -> HashMap()}
  }

  fun putSetting(user : User, settingName : String, value : String) : SettingEnum {
    return when {
      userChangeable[settingName] == null -> SettingEnum.DOESNOTEXIST
      userChangeable[settingName]!! -> putSetting(user.id, settingName, value)
      else -> SettingEnum.FORBIDDEN
    }
  }

  fun putSetting(guild : Guild, settingName: String, value: String) : SettingEnum {
    return when {
      userChangeable[settingName] == null -> SettingEnum.DOESNOTEXIST
      userChangeable[settingName]!! -> putSetting(guild.id, settingName, value)
      else -> SettingEnum.FORBIDDEN
    }
  }

  fun putSetting(settingName: String, value: String) : SettingEnum {
    return when {
      defaultSettings[settingName] == null -> SettingEnum.DOESNOTEXIST
      else -> putSetting("DEFAULT", settingName, value)
    }
  }

  private fun putSetting(owner : String, settingName: String, value: String) : SettingEnum {
    println("Changing setting $settingName for $owner to $value")
    return try {
      customSettings[settingName]?.set(owner, value)
      writeJSON(settingName, owner, value)
      SettingEnum.SUCCESSFUL
    } catch (_ : Exception) {
      SettingEnum.ERROR
    }
  }

  fun removeSetting(user : User, settingName: String) : SettingEnum {
    return when {
      userChangeable[settingName] == null -> SettingEnum.DOESNOTEXIST
      userChangeable[settingName]!! -> removeSetting(user.id, settingName)
      else -> SettingEnum.FORBIDDEN
    }
  }

  fun removeSetting(guild: Guild, settingName: String) : SettingEnum {
    return when {
      userChangeable[settingName] == null -> SettingEnum.DOESNOTEXIST
      userChangeable[settingName]!! -> removeSetting(guild.id, settingName)
      else -> SettingEnum.FORBIDDEN
    }
  }

  fun removeSetting(settingName: String) : SettingEnum {
    return when {
      defaultSettings[settingName] == null -> SettingEnum.DOESNOTEXIST
      else -> removeSetting("DEFAULT", settingName)
    }
  }

  private fun removeSetting(owner : String, settingName: String) : SettingEnum {
    println("Removing setting $settingName for $owner.")
    if (!customSettings[settingName]?.containsKey(owner)!!) {
      println("Nevermind it wasn't set")
      return SettingEnum.NOTSET
    }

    return try {
      customSettings[settingName]?.remove(owner)
      writeJSON(settingName, owner, null)
      SettingEnum.SUCCESSFUL
    } catch (_ : Exception) {
      SettingEnum.ERROR
    }
  }

  fun getSetting(settingName: String, message: Message) : String {
    val defaultSetting = defaultSettings[settingName] ?: "This response is broken :("
    val setting = customSettings[settingName] ?: return standardReplacements(defaultSetting, message)

    // Return user setting if we have one
    val userSetting = setting[message.author.id]
    if (userSetting != null) { return standardReplacements(userSetting, message) }

    // Return guild setting if message is from guild and we have one
    if (message.isFromGuild) {
      val guildSetting = setting[message.guild.id]
      if (guildSetting != null) { return standardReplacements(guildSetting, message) }
    }

    // Return altered default value if we have one
    val customSetting = setting["DEFAULT"]
    if (customSetting != null) { return standardReplacements(customSetting, message) }

    // Fall back to returning the default value if we can't find any of the above
    return standardReplacements(defaultSetting, message)
  }

  private fun standardReplacements(input : String, message: Message) : String {
    return input
      .replace("\\{(user|author)\\}".toRegex(), message.author.asMention)
      .replace("\\{(bot|botname|coffeebro)\\}".toRegex(), message.jda.selfUser.asMention)
      .replace("\\{(message|content)\\}".toRegex(), message.contentRaw.trim())
  }

  fun allSettingsContaining(substring : String) : Set<String> {
    return customSettings
      .keys
      .stream()
      .filter { x -> x.contains(substring) }
      .collect(Collectors.toSet())
  }
}